File: Java/Structures/Tree.java
package Structures;
class Node{
    int data;
    Node left = null;
    Node right = null;
    public Node(int data){
        this.data = data;
    }
    public Node(){}
}

public class Tree {
    public Node root = null;
    public Tree(){}
    
    public void displayInorder(Node ptr){
        if(ptr!=null){
            displayInorder(ptr.left);
            System.out.println(ptr.data);
            displayInorder(ptr.right);
        }
    }

    public void displayPreorder(Node ptr){
        if(ptr!=null){
            System.out.println(ptr.data);
            displayPreorder(ptr.left);
            displayPreorder(ptr.right);
        }
    }

    public void displayPostorder(Node ptr){
        if(ptr!=null){
            displayPostorder(ptr.left);
            displayPostorder(ptr.right);
            System.out.println(ptr.data);
        }
    }

    public void insertNode(Node node){
        insertNodeElement(node.data);
    }

    public void insertNodeElement(int data){
        Node node = new Node(data);
        if(is_empty()){
            root = node;
        }else{
            Node ptr = root;
            while(ptr.left != null || ptr.right != null){
                if(node.data < ptr.data)
                    ptr = ptr.left;
                if(node.data > ptr.data)
                    ptr = ptr.right;
                if(node.data == ptr.data)
                    return; // no duplicate
            }

            if(ptr.data < node.data)
                ptr.right = node;
            else
                ptr.left = node;
        }
    }

    public Node deleteNode(Node ptr, int data){
        if(ptr==null)
            return ptr;
        else if(data < ptr.data)
            ptr.left = deleteNode(ptr.left, data);
        else if(data > ptr.data)
            ptr.right = deleteNode(ptr.right, data);
        else{
            // case 1: leaf node
            if(ptr.left == null && ptr.right == null){
                ptr = null;
            }
            // case 2: one child
            else if(ptr.left == null){
                Node temp = ptr;
                ptr = ptr.right;
                temp = null;
            }
            else if(ptr.right == null){
                Node temp = ptr;
                ptr = ptr.left;
                temp = null;
            }
            // case 3: 2 children
            else{
                Node temp  = findMin(ptr.right);
                ptr.data = temp.data;
                ptr.right = deleteNode(ptr.right, temp.data);
            }

            return ptr;
        }

    }

    public Node findMin(Node ptr){
        while(ptr.left != null)
            ptr = ptr.left;
        return ptr;
    }

    public Node findMax(Node ptr){
        while(ptr.right != null)
            ptr = ptr.right;
        return ptr;
    }

    public boolean is_element(int data){
        Node ptr = root;
        while(ptr.left != null || ptr.right != null){
            if(data < ptr.data)
                ptr = ptr.left;
            if(data > ptr.data)
                ptr = ptr.right;
            if(data == ptr.data)
                return true; 
        }
        return false;
    }

    private boolean is_empty(){
        return root == null;
    }
}

class MainClass{
    public static void main(String[] args){
        Tree tree = new Tree();

        for(int i = 0; i < 10; ++i){
            tree.insertNodeElement(i);
        }
        tree.displayPostorder(tree.root);
    }
}

File: Python/Structures/Tree.py
class Node:
    def __init__(self, data=None):
        self.left = None
        self.right = None
        self.data = data


class Tree:
    m_size: int = 0

    def __init__(self):
        self.m_root = None

    def insert_node(self, node: Node) -> None:
        if self.is_empty():
            self.m_root = node
        else:
            ptr = self.m_root
            while ptr != None:
                if node.data < ptr.data:
                    ptr = ptr.left
                elif node.data > ptr.data:
                    ptr = ptr.right
                else:
                    return

            if node.data < ptr.data:
                ptr.left = node
            if node.data > ptr.data:
                ptr.right = node

        self.m_size += 1

    def insert_element(self, data: int) -> None:
        node = Node(data)
        if self.is_empty():
            self.m_root = node
        else:
            ptr = self.m_root
            while ptr.left is not None or ptr.right is not None:
                if node.data < ptr.data:
                    ptr = ptr.left
                elif node.data > ptr.data:
                    ptr = ptr.right
                else:
                    return

            if node.data < ptr.data:
                ptr.left = node
            if node.data > ptr.data:
                ptr.right = node

        self.m_size += 1

    def traverse_inorder(self, node: Node) -> None:
        if node is not None:
            self.traverse_inorder(node.left)
            print(node.data)
            self.traverse_inorder(node.right)

    def traverse_preorder(self, node: Node) -> None:
        if node is not None:
            print(node.data)
            self.traverse_preorder(node.left)
            self.traverse_preorder(node.right)

    def traverse_postorder(self, node: Node) -> None:
        if node is not None:
            self.traverse_postorder(node.left)
            self.traverse_postorder(node.right)
            print(node.data)

    def traverse_levelorder(self, node: Node) -> None:
        pass

    def delete_element(self, node: Node, data: int) -> typing.Optional[Node]:
        if node is None:
            return None
        elif data < node.data:
            node.left = self.delete_element(node.left, data)
        elif data > node.data:
            node.right = self.delete_element(node.right, data)
        else:
            # case 1:
            if node.left is None and node.right is None:
                ptr = None
            # case 2:
            elif node.left is None:
                ptr = node
                node = node.right
                ptr = None

            elif node.right is None:
                ptr = node
                node = node.left
                ptr = None
            # case 3:
            else:
                ptr = self.find_min(node)
                node.data = ptr.data
                ptr.right = self.delete_element(node.right, ptr.data)

    def find_min(self, node: Node) -> Node:
        while node.left is not None:
            node = node.left
        return node

    def is_empty(self) -> bool:
        return self.m_root == None


if __name__ == "__main__":
    t: Tree = Tree()
    ptr: Node = t.m_root

    for i in range(0, 10):
        t.insert_element(i)

    t.traverse_postorder(t.m_root)
    print("done")
